lnwire/signature_test.go

```diff
@@ -24,67 +25,80 @@ func TestSignatureSerializeDeserialize(t *testing.T) {
 			return err
 		}
 
-		if e.R.Cmp(e2.R) != 0 {
-			return fmt.Errorf("Pre/post-serialize Rs don't match"+
-				": %s, %s", e.R, e2.R)
-		}
-		if e.S.Cmp(e2.S) != 0 {
-			return fmt.Errorf("Pre/post-serialize Ss don't match"+
-				": %s, %s", e.S, e2.S)
+		if !e.IsEqual(e2) {
+			return fmt.Errorf("pre/post-serialize sigs don't "+
+				"match: %x, %x", e.Serialize(), e2.Serialize())
 		}
 		return nil
 	}
 
-	sig := btcec.Signature{}
-
 	// Check R = N-1, S = 128.
-	sig.R = big.NewInt(1) // Allocate a big.Int before we call .Sub.
-	sig.R.Sub(btcec.S256().N, sig.R)
-	sig.S = big.NewInt(128)
+	r := big.NewInt(1) // Allocate a big.Int before we call .Sub.
+	r.Sub(btcec.S256().N, r)
+	rScalar := new(btcec.ModNScalar)
+	rScalar.SetByteSlice(r.Bytes())
+
+	sig := ecdsa.NewSignature(rScalar, new(btcec.ModNScalar).SetInt(128))
 	err := signatureSerializeDeserialize(sig)
 	if err != nil {
 		t.Fatalf("R = N-1, S = 128: %s", err.Error())
 	}
 
 	// Check R = N-1, S = 127.
-	sig.S = big.NewInt(127)
+	sig = ecdsa.NewSignature(rScalar, new(btcec.ModNScalar).SetInt(127))
 	err = signatureSerializeDeserialize(sig)
 	if err != nil {
 		t.Fatalf("R = N-1, S = 127: %s", err.Error())
 	}
 
 	// Check R = N-1, S = N>>1.
-	sig.S.Set(btcec.S256().N)
-	sig.S.Rsh(sig.S, 1)
+	s := new(big.Int).Set(btcec.S256().N)
+	s.Rsh(s, 1)
+	sScalar := new(btcec.ModNScalar)
+	sScalar.SetByteSlice(s.Bytes())
+	sig = ecdsa.NewSignature(rScalar, sScalar)
 	err = signatureSerializeDeserialize(sig)
 	if err != nil {
 		t.Fatalf("R = N-1, S = N>>1: %s", err.Error())
 	}
 
 	// Check R = N-1, S = N.
-	sig.S.Set(btcec.S256().N)
+	s = new(big.Int).Set(btcec.S256().N)
+	overflow := sScalar.SetByteSlice(s.Bytes())
+	if !overflow {
+		t.Fatalf("Expect ModNScalar to overflow when setting N but " +
+			"didn't")
+	}
+	sig = ecdsa.NewSignature(rScalar, sScalar)
 	err = signatureSerializeDeserialize(sig)
-	if err.Error() != "signature S isn't 1 or more" {
+	if err.Error() != "invalid signature: S is 0" {
 		t.Fatalf("R = N-1, S = N should become R = N-1, S = 0: %s",
 			err.Error())
 	}
 
 	// Check R = N-1, S = N-1.
-	sig.S.Sub(sig.S, big.NewInt(1))
-	err = signatureSerializeDeserialize(sig)
-	if err.Error() != "Pre/post-serialize Ss don't match: 115792089237316"+
-		"195423570985008687907852837564279074904382605163141518161494"+
-		"336, 1" {
-		t.Fatalf("R = N-1, S = N-1 should become R = N-1, S = 1: %s",
-			err.Error())
-	}
+	//s = new(big.Int).Set(btcec.S256().N)
+	//s.Sub(s, big.NewInt(1))
+	//sScalar.SetByteSlice(s.Bytes())
+	//sig = ecdsa.NewSignature(rScalar, sScalar)
+	//err = signatureSerializeDeserialize(sig)
+	//if err.Error() != "Pre/post-serialize Ss don't match: 115792089237316"+
```

Comment:
Something seems to be serialized differently here. @Roasbeef, not sure if I did anything wrong here?

Reply:
I wasn't able to re-create this test case fully, at least not in regard to its error message. The problem is: the sigs don't match (as expected) because `Serialize()` correctly fixes/unifies the low-s value. But we can't access the internal state to show the difference, we can only call `Serialize()` to give us the raw bytes (which again fix the low-s value).
So tl;dr is: we can only assert that the sig is different but cannot show the actual difference of its internal state (unless we do some nasty reflection magic.

Reply:
Ah makes sense, the `Serialize()` we use `Serialize()` we use is the `dcrec` version that does indeed fix those issues as part of encoding: https://github.com/decred/dcrd/blob/dcrec/secp256k1/v4.0.1/dcrec/secp256k1/ecdsa/signature.gao#L89-L96. The prior one does as well, but this one also attempts to more aggressively strip out leading zeros. 

Nice to tie up this loose end to make sure it doesn't bit us in the future!